File: vane-core/src/main/java/org/oddlama/vane/core/config/loot/LootTableList.java
package org.oddlama.vane.core.config.loot;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.oddlama.vane.core.config.ConfigDictSerializable;

public class LootTableList implements ConfigDictSerializable {

    private List<LootDefinition> tables = new ArrayList<>();

    public List<LootDefinition> tables() {
        return tables;
    }

    // Map special internal names to PascalCase names for YAML generation
    public Map<String, Object> toDict() {
        return tables.stream().collect(Collectors.toMap(t -> toYamlName(t.name()), LootDefinition::serialize));
    }

    public void fromDict(final Map<String, Object> dict) {
        tables.clear();
        for (final Map.Entry<String, Object> e : dict.entrySet()) {
            final String name = e.getKey();
            final var internalName = name == null || name.isEmpty() ? name : fromYamlName(name);
            tables.add(LootDefinition.deserialize(internalName, e.getValue()));
        }
    }

    public static LootTableList of(LootDefinition... defs) {
        final var rl = new LootTableList();
        rl.tables = Arrays.asList(defs);
        return rl;
    }

    private static String toYamlName(final String s) {
        if (s == null || s.isEmpty()) return s;
        final var low = s.toLowerCase();
        switch (low) {
            case "generic":
                return "Generic";
            case "terralith_generic":
                return "TerralithGeneric";
            case "terralith_rare":
                return "TerralithRare";
            case "ancientcity":
                return "AncientCity";
            case "bastion":
                return "Bastion";
            default:
                return s;
        }
    }

    private static String fromYamlName(final String s) {
        if (s == null || s.isEmpty()) return s;
        final var low = s.toLowerCase().replaceAll("[_\\s]", "");
        switch (low) {
            case "generic":
                return "generic";
            case "terralithgeneric":
                return "terralith_generic";
            case "terralithrare":
                return "terralith_rare";
            case "ancientcity":
                return "ancientcity";
            case "bastion":
                return "bastion";
            default:
                return s;
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/loot/LootTableList.kt
package org.oddlama.vane.core.config.loot

import org.oddlama.vane.core.config.ConfigDictSerializable
import java.util.Locale

class LootTableList : ConfigDictSerializable {
    private var tables: MutableList<LootDefinition?> = mutableListOf()

    fun tables(): MutableList<LootDefinition?> = tables

    // Map special internal names to PascalCase names for YAML generation
    override fun toDict(): MutableMap<String, Any> =
        tables.filterNotNull().associateTo(mutableMapOf()) { t -> toYamlName(t.name)!! to (t.serialize() as Any) }

    override fun fromDict(dict: MutableMap<String, Any>) {
        tables.clear()
        dict.entries.forEach { (name, raw) ->
            val internalName = if (name.isEmpty()) name else fromYamlName(name)
            tables.add(LootDefinition.deserialize(internalName, raw))
        }
    }

    companion object {
        @JvmStatic
        fun of(vararg defs: LootDefinition?): LootTableList =
            LootTableList().also { it.tables = defs.toMutableList() }

        private fun toYamlName(s: String?): String? {
            if (s.isNullOrEmpty()) return s
            return when (s.lowercase(Locale.getDefault())) {
                "generic"          -> "Generic"
                "terralith_generic"-> "TerralithGeneric"
                "terralith_rare"   -> "TerralithRare"
                "ancientcity"      -> "AncientCity"
                "bastion"          -> "Bastion"
                else               -> s
            }
        }

        private fun fromYamlName(s: String?): String? {
            if (s.isNullOrEmpty()) return s
            return when (s.lowercase(Locale.getDefault()).replace("[_\\s]".toRegex(), "")) {
                "generic"          -> "generic"
                "terralithgeneric" -> "terralith_generic"
                "terralithrare"    -> "terralith_rare"
                "ancientcity"      -> "ancientcity"
                "bastion"          -> "bastion"
                else               -> s
            }
        }
    }
}
